cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/WorkflowStatsSLAHttpHandler.java

```diff
@@ -89,4 +111,197 @@ public void workflowStats(HttpRequest request, HttpResponder responder,
+          workflowRunMetrics.getWorkflowRunId(), programMetrics.getProgramRunId(),
+          programMetrics.getProgramStartTime(), programMetrics.getMetrics());
+      }
+    }
+    return programLevelDetailsMap.values();
+  }
+
+  /**
+   * Returns the detailed Record for the Workflow
+   *
+   * @param workflowId Workflow that needs to get its detailed record
+   * @param runId Run Id of the workflow
+   * @return Return the Workflow Run Metrics
+   * @throws Exception
+   */
+  @Nullable
+  private WorkflowRunMetrics getDetailedRecord(Id.Workflow workflowId, String runId) throws Exception {
+    WorkflowDataset.WorkflowRunRecord workflowRunRecord = store.getWorkflowRun(workflowId, runId);
+    if (workflowRunRecord == null) {
+      return null;
+    }
+    List<WorkflowDataset.ProgramRun> programRuns = workflowRunRecord.getProgramRuns();
+    List<ProgramMetrics> programMetricsList = new ArrayList<>();
+    for (WorkflowDataset.ProgramRun programRun : programRuns) {
+      Map<String, Long> programMap = new HashMap<>();
+      Id.Program program = Id.Program.from(workflowId.getNamespaceId(), workflowId.getApplicationId(),
+                                           programRun.getProgramType(), programRun.getName());
+      if (programRun.getProgramType() == ProgramType.MAPREDUCE) {
+        programMap = getMapreduceDetails(program, programRun.getRunId());
+      } else if (programRun.getProgramType() == ProgramType.SPARK) {
+        programMap = getSparkDetails(program, programRun.getRunId());
+      }
+      programMap.put("timeTaken", programRun.getTimeTaken());
+      programMetricsList.add(new ProgramMetrics(
+        programRun.getName(), programRun.getProgramType(), programRun.getRunId(),
+        RunIds.getTime(RunIds.fromString(programRun.getRunId()), TimeUnit.SECONDS), programMap));
+    }
+    return new WorkflowRunMetrics(runId, programMetricsList);
+  }
+
+  private Map<String, Long> getMapreduceDetails(Id.Program mapreduceProgram, String runId) throws Exception {
+    Id.Run mrRun = new Id.Run(mapreduceProgram, runId);
+    return mrJobInfoFetcher.getMRJobInfo(mrRun).getCounters();
+  }
+
+  private Map<String, Long> getSparkDetails(Id.Program sparkProgram, String runId) throws Exception {
+    Map<String, String> context = new HashMap<>();
+    context.put(Constants.Metrics.Tag.NAMESPACE, sparkProgram.getNamespaceId());
+    context.put(Constants.Metrics.Tag.APP, sparkProgram.getApplicationId());
+    context.put(Constants.Metrics.Tag.SPARK, sparkProgram.getId());
+    context.put(Constants.Metrics.Tag.RUN_ID, runId);
+
+    List<TagValue> tags = new ArrayList<>();
+    for (Map.Entry<String, String> entry : context.entrySet()) {
+      tags.add(new TagValue(entry.getKey(), entry.getValue()));
+    }
+    MetricSearchQuery metricSearchQuery = new MetricSearchQuery(0, 0, Integer.MAX_VALUE, tags);
+    Collection<String> metricNames = metricStore.findMetricNames(metricSearchQuery);
```

Comment:
@gsps1
Shouldn't the findMetricNames not require a MetricSearchQuery object (which has start time, end time, and limit - unnecessary for finding metric names)?